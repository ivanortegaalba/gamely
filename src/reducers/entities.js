import { get } from 'lodash'

const initialEntities = {
  games: []
}
function entities (state = initialEntities, action) {
  if (action.response && action.response.entities) {
    let newState = Object.assign({}, state)
    for (let entity in action.response.entities) {
      if (action.response.entities.hasOwnProperty(entity)) {
        newState[entity] = {
          ...action.response.entities[entity]
        }
      }
    }
    return newState
  }
  return state
}

function actionNames (entityName) {
  entityName = entityName.toUpperCase()
  const requestType = entityName + '_REQUEST'
  const successType = entityName + '_SUCCESS'
  const failureType = entityName + '_FAILURE'

  return [requestType, successType, failureType]
}

export function fetchEntity ({ entityName, isList = true, isNormalized = true }) {
  let requestType
  let successType
  let failureType
  ;[requestType, successType, failureType] = actionNames(entityName)

  let initialState = { isFetching: false, error: false }

  if (isList === true) {
    initialState['items'] = []
  }
  return (state = initialState, action) => {
    switch (action.type) {
      case requestType:
        return {
          ...state,
          isFetching: true
        }
      case successType: {
        return {
          isFetching: false,
          error: false,
          items: action.response.result
        }
      }
      case failureType:
        return {
          ...state,
          isFetching: false,
          error: true
        }
      default:
        return state
    }
  }
}

export default entities
