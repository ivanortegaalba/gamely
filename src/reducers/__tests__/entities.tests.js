import gameMock from '../../__mocks__/game';
import entities, { actionNames, reducers } from '../entities';

describe('entities reducers', () => {
  describe('entities', () => {
    it('should provide the initial state', () => {
      expect(entities(undefined, {})).toEqual({
        games: [],
      });
    });

    describe('when a normalized response is received', () => {
      const response = {
        result: [1],
        entities: {
          games: {
            1: gameMock,
          },
        },
      };
      it('should handle ANY action with response as data', () => {
        expect(
          entities(undefined, {
            type: 'WHATEVER_ACTION',
            response,
          })
        ).toEqual(response.entities);
      });

      it('should overwrite the entity with the same ID', () => {
        const initialEntities = {
          games: {
            1: {},
          },
        };
        expect(
          entities(initialEntities, {
            type: 'WHATEVER_ACTION',
            response,
          })
        ).toEqual(response.entities);
      });
    });
  });

  describe('actionNames', () => {
    it('given a entity name, generate an array with each entity action type', () => {
      expect(actionNames('GAME')).toEqual([
        'GAME_REQUEST',
        'GAME_SUCCESS',
        'GAME_FAILURE',
      ]);
    });
  });

  describe('fetchEntity', () => {
    const { fetchEntity } = reducers;
    const response = {
      result: [1],
      entities: {
        games: {
          1: gameMock,
        },
      },
    };
    const actionTypes = {
      REQUEST: {
        type: 'GAME_REQUEST',
      },
      FAILURE: {
        type: 'GAME_FAILURE',
      },
      SUCCESS: {
        type: 'GAME_SUCCESS',
        response,
      },
    };
    Object.keys(actionTypes).forEach(actionType => {
      describe(`for *_${actionType} type`, () => {
        const action = actionTypes[actionType];
        it('should provide the initial state', () => {
          expect(fetchEntity('GAME')(undefined, action)).toMatchSnapshot();
        });
      });
    });
  });
});
