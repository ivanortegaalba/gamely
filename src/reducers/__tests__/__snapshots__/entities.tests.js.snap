// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entities reducers fetchEntity for *_FAILURE type should provide the initial state 1`] = `
Object {
  "error": true,
  "isFetching": false,
  "items": Array [],
}
`;

exports[`entities reducers fetchEntity for *_REQUEST type should provide the initial state 1`] = `
Object {
  "error": false,
  "isFetching": true,
  "items": Array [],
}
`;

exports[`entities reducers fetchEntity for *_SUCCESS type should provide the initial state 1`] = `
Object {
  "error": false,
  "isFetching": false,
  "items": Array [
    1,
  ],
}
`;
